src/internal/pjs/server_internal_test.go

```diff
@@ -0,0 +1,480 @@
+		s := status.Convert(err)
+		require.Equal(t, codes.NotFound, s.Code())
+	})
+}
+
+func TestRunJob(t *testing.T) {
+	c, fc := setupTest(t)
+	ctx := pctx.TestContext(t)
+	programFileset := createFileSet(t, fc, map[string][]byte{
+		"file": []byte(`!#/bin/bash; ls /input/;`),
+	})
+	inputFileset := createFileSet(t, fc, map[string][]byte{
+		"a.txt": []byte("dummy input"),
+	})
+	in := &pjs.CreateJobRequest{
+		Program: programFileset,
+		Input:   []string{inputFileset},
+	}
+
+	out, err := runJob(t, ctx, c, in, func(resp *pjs.ProcessQueueResponse) error {
+		// todo(muyang)
+		// for now we do nothing here, simply return the input filesets
+		resp.Input = in.Input
+		return nil
+	})
+	require.NoError(t, err)
+	require.Equal(t, in.Input, out.Output)
+}
+
+func TestCancelJob(t *testing.T) {
+	t.Run("cancel job before processing", func(t *testing.T) {
+		ctx := pctx.TestContext(t)
+		c, fc := setupTest(t)
+		resp := createJob(t, ctx, c, fc)
+		jid := resp.Id.Id
+		_, err := c.CancelJob(ctx, &pjs.CancelJobRequest{
+			Job: &pjs.Job{
+				Id: jid,
+			},
+		})
+		require.NoError(t, err)
+		inspectJobResp, err := c.InspectJob(ctx, &pjs.InspectJobRequest{
+			Job: &pjs.Job{
+				Id: jid,
+			},
+		})
+		require.NoError(t, err)
+		require.NotNil(t, inspectJobResp.Details)
+		jobInfo := inspectJobResp.Details.JobInfo
+		require.Equal(t, pjs.JobState_DONE, jobInfo.State)
+		// todo: the queue should be empty. This can be tested when inspectQueue is ready
+	})
+	t.Run("cancel a processing job", func(t *testing.T) {
+		ctx := pctx.TestContext(t)
+		c, fc := setupTest(t)
+
+		program := createFileSet(t, fc, map[string][]byte{"program": []byte("foo")})
+		createResp := createJob(t, ctx, c, fc, func(req *pjs.CreateJobRequest) {
+			req.Program = program
+		})
+
+		cancel := func() {
+			_, err := c.CancelJob(ctx, &pjs.CancelJobRequest{
+				Job: &pjs.Job{
+					Id: createResp.Id.Id,
+				},
+			})
+			require.NoError(t, err)
+		}
+
+		_, err := runJobFrom(t, ctx, c, createResp.Id.Id, program, func(resp *pjs.ProcessQueueResponse) error {
+			cancel()
+			// simulate processing time and ensure cancelJob is finished
+			time.Sleep(5 * time.Second)
+			return nil
+		})
+		require.NoError(t, err)
+
+		inspectCancelledJob(t, ctx, c, createResp.Id.Id)
+	})
+	t.Run("cancel processing job tree", func(t *testing.T) {
+		ctx := pctx.TestContext(t)
+		c, fc := setupTest(t)
+		eg, egCtx := errgroup.WithContext(ctx)
+
+		jobAProcessing := make(chan struct{})
+		jobBProcessing := make(chan struct{})
+		jobCProcessing := make(chan struct{})
+
+		var jobAContext string
+		programA := createFileSet(t, fc, map[string][]byte{"program": []byte("A")})
+		createRespA := createJob(t, ctx, c, fc, func(req *pjs.CreateJobRequest) {
+			req.Program = programA
+		})
+		eg.Go(func() error {
+			_, err := runJobFrom(t, egCtx, c, createRespA.Id.Id, programA, func(resp *pjs.ProcessQueueResponse) error {
+				jobAContext = resp.Context
+				close(jobAProcessing)
+				time.Sleep(5 * time.Second)
```

Comment:
I don't think you want to explicitly wait for this long.  It would be better to have another "jobsDone" or "jobADone" channel that you close, so that you can unblock when you're ready.  I think this probably delays CancelJob succeeding as well; you might want:

```go
select {
case <-time.After(5*time.Second):
    return errors.New("job ran too long")
case <-ctx.Done():
    return context.Cause(ctx)
}
```
This lets the timer stop as soon as the context is done, which is what I'm guessing happens when a running job is canceled.